internal/pkg/providers/azure/azure.go

```diff
@@ -0,0 +1,238 @@
+	} else {
+		logrus.Error("filter type " + filter.FilterType.String() + " not supported for azure yet")
+		return nil, errors.New("filter type " + filter.FilterType.String() + " not supported for azure yet")
+	}
+}
+
+func fetchResourceGroupsByFilter(azure *Azure, filter config.Filter) (result *resources.ListResultPage, err error) {
+	resourcesClient, err := getResourcesClient(azure)
+	if err != nil {
+		return nil, err
+	}
+	tagFilter := fmt.Sprintf("tagName eq '%s' and tagValue eq '%s'", filter.LabelName, filter.LabelValue)
+	resourceGroups, err := resourcesClient.ListByResourceGroup(context.TODO(), azure.ResourceGroupName, tagFilter, "SecurityRules", nil)
+	if err != nil {
+		return nil, err
+	}
+	return &resourceGroups, nil
+}
+
+func fetchSecurityGroup(a *Azure, resourceName string) (securityGroup network.SecurityGroup, err error) {
+	securityGroupClient, err := getSecurityGroupsClient(a)
+	if err != nil {
+		return network.SecurityGroup{}, err
+	}
+	securityGroup, err = securityGroupClient.Get(context.TODO(), a.ResourceGroupName, resourceName, "")
+	if err != nil {
+		return network.SecurityGroup{}, err
+	}
+	return securityGroup, err
+}
+
+func getSecurityRulesClient(a *Azure) (network.SecurityRulesClient, error) {
+
+	authorizer, err := auth.NewClientCredentialsConfig(a.ClientID, a.ClientSecret, a.TenantID).Authorizer()
+	if err != nil {
+		return network.SecurityRulesClient{}, err
+	}
+	securityRulesClient := network.NewSecurityRulesClient(a.SubscriptionID)
+	securityRulesClient.Authorizer = authorizer
+
+	return securityRulesClient, nil
+}
+
+func deleteRule(rulesClient network.SecurityRulesClient, a *Azure, resourceName string, ipDescription string) error {
+	futureT, err := rulesClient.Delete(context.TODO(), a.ResourceGroupName, resourceName, ipDescription)
+	if err != nil {
+		return err
+	}
+	err = futureT.WaitForCompletionRef(context.TODO(), rulesClient.Client)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func getResourcesClient(a *Azure) (resources.Client, error) {
+
+	authorizer, err := auth.NewClientCredentialsConfig(a.ClientID, a.ClientSecret, a.TenantID).Authorizer()
+	if err != nil {
+		return resources.Client{}, err
+	}
+	resourcesClient := resources.NewClient(a.SubscriptionID)
+	resourcesClient.Authorizer = authorizer
+
+	return resourcesClient, nil
+}
+
+func getSecurityGroupsClient(a *Azure) (*network.SecurityGroupsClient, error) {
+
+	authorizer, err := auth.NewClientCredentialsConfig(a.ClientID, a.ClientSecret, a.TenantID).Authorizer()
+	if err != nil {
+		return nil, err
+	}
+	nsgClient := network.NewSecurityGroupsClient(a.SubscriptionID)
+	nsgClient.Authorizer = authorizer
+
+	return &nsgClient, nil
+}
+
+func createSecurityRule(a *Azure, rulesClient network.SecurityRulesClient, resourceName string, ipDescription *string, fromPortStr string, toPortStr string, ipRange *string, ipProtocol string) error {
+
+	var networkProtocol network.SecurityRuleProtocol
+
+	switch ipProtocol {
```

Comment:
Perhaps move this to a new method like `validateSecurityRuleProtocol` to keep this method small.